CMake: overhaul and improve scripts

FranciscoPombal opened the pull request:
@glassez @Chocobo1 @eli-schwartz 

This is not meant to supersede https://github.com/qbittorrent/qBittorrent/pull/12746 (at least not right off the bat ...). It is meant to provide a comparison for it. It is built directly on top of it, to enable easily studying the differences from the diff, both against master and the genexp alternative.

I gave it the best possible shot I could for this alternative to "look good". The logic is meant to be the exact same as https://github.com/qbittorrent/qBittorrent/pull/12746, but with no generator expressions.  Any errors/incorrect/suboptimal things are honest mistakes (I'm pretty sure I got everything right, though). I tried to follow the structure outlined in https://github.com/qbittorrent/qBittorrent/pull/12746#issuecomment-687137672 as much as possible:

>So I think it's much more convenient to see the following CMake project structure (generally):
>1. Common declarations 
>2. Platform dependent declarations (per platform)
>3. Feature dependent declarations (per feature)
>3.1. Common declarations 
>3.2. Platform dependent declarations (per platform)

There were only 2 generator expressions left that I could not avoid. I left a comment starting with `# GENEXP:` in the corresponding locations.

Feel free to post your opinions. Personally I think this is a step backwards, for the reasons I already explained before.

Note: there is a small amount of noise stemming from a trivial change from `MATCHES` to `STREQUAL` here and there. Sorry about that.

glassez commented on CMakeLists.txt:
```diff
@@ -1,71 +1,52 @@
+)
 
+# use CONFIG mode first in find_package
+set(CMAKE_FIND_PACKAGE_PREFER_CONFIG ON)
```
This and related changes (removal of FindLibtorrent file) seems quite trivial and independent so you can extract it into separate PR.

FranciscoPombal replied:
@glassez I don't think so. These are some of the more important changes in this PR.

glassez replied:
>These are some of the most important changes in this PR

...that doesn't prevent it from separating from everything else.
Let me remind you again that my intention is to separate the easily understandable (separate) parts so that we can quickly review and merge them, which should simplify the reviewing of the remaining part. It really takes too much time to review all current PR every time you change it.

FranciscoPombal replied:
@glassez 

> It really takes too much time to review all current PR every time you change it.

This not something that I have changed since the previous PR. In fact, this part has not changed since before I unmarked the other one as draft. This change in how libtorrent is found has always been one of the primary motivations for this whole thing. Previously it was too complex and relied on a self-bundled find module that required maintenance, lest it become to outdated and cause problems.

Now it is really simple:

- If not on "Unix", find libtorrent as a `REQUIRED` package just like every other package that's mandatory for compilation.
- If on "Unix", first find the package as non-`REQUIRED`, so that if it fails, there is the second chance of finding it with pkg-config.
    - In either case, set its properties indicate that it's a `REQUIRED` package so that it shows up correctly in the feature summary output in the list of `REQUIRED` packages.

`CMAKE_FIND_PACKAGE_PREFER_CONFIG` is set to on so that CMake first tries to find packages by CMake config files, before trying to find them via `Find<Pkg>` modules. If a distro maintainer builds libtorrent in such a way that it doesn't provide CMake config files, and pkg-config is not desirable/available, they can still bundle their custom `FindLibtorrent.cmake` module in the `cmake/Modules` directory, and CMake will use that.
Finding packages via CMake config files supplied by the libraries themselves is the preferable way. For example, this is how Qt5 packages are found.

glassez replied:
@FranciscoPombal, are you kidding me by writing such long comments? 🤔
You didn't add anything new to what was said about it earlier.
But nevertheless, after reading this comment even for the first time, someone realizes that these changes are independent enough to become a separate PR. However, as you wish... if this insistence of yours increases the review time (there will be at least one other reviewer who is most likely not so familiar with the content of these changes), you are to blame.

FranciscoPombal replied:
@glassez 

> @FranciscoPombal, are you kidding me by writing such long comments? thinking
> You didn't add anything new to what was said about it earlier.
> But nevertheless, after reading this comment even for the first time, someone realizes that these changes are independent enough to become a separate PR. However, as you wish... if this insistence of yours increases the review time (there will be at least one other reviewer who is most likely not so familiar with the content of these changes), you are to blame.

I'm just trying to explain as clearly as possible. I'm confused as to why this is posing a problem _now_, since it had previously been solved after extensive discussion.

I have already stated that I agree with separating into separate commits after the initial review is finished, but separating essential parts of the PR into separate PRs doesn't make sense.

glassez commented on CMakeLists.txt:
```diff
@@ -1,71 +1,52 @@
+# features (some are platform-specific)
+include(FeatureSummary)
+include(FeatureOptionsSetup)
+feature_option(BUILD_SHARED_LIBS "Build internal libraries SHARED instead of STATIC (for experimental development purposes)" OFF)
```
I don't remember it being present the last time I reviewed it (or I didn't notice it because of other issues).
It is very questionable (as well as current splitting into several libraries), so I would remove it before it becomes a new stumbling block here.

FranciscoPombal replied:
@glassez 

`BUILD_SHARED_LIBS` is a built-in CMake construct. It makes every "naked" `add_library()` call (i.ie., one that **does not** specify `STATIC` or `SHARED`) act as `STATIC` or `SHARED`, depending on whether it is `OFF` (default) or `ON`, respectively. Wrapping it in inside `feature_option()` is just a convenience to have its description printed out - it makes sense to me for it to be printed, since it provides such important functionality and is already there.

When it's off (the default) it doesn't change anything that we were already doing.

Declaring libraries with "naked" `add_library()` is good for flexibility, because it gives the builder the choice via the simple meachanism of passing `-DBUILD_SHARED_LIBS=ON/OFF`. Currently, qBittorrent does not build at all with `-DBUILD_SHARED_LIBS=ON` (hence the note about "experimental development purposes"), but this may change in the future, unless we decide that there is no `add_library()` call in our codebase that would ever make sense to offer such flexibility.

Docs: https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html

glassez replied:
Currently qBittorrent isn't supposed to be a set of several libraries.
The fact that CMake project divides it into several (static) libraries can only be considered as a workaround that allows CMake to follow the file structure of the project's source code (perhaps due to some drawbacks that prevent it from being done in another way).
You should not change this here, especially since the build as a set of shared libraries is not possible currently (at least on Windows, it requires changes to the source code).

FranciscoPombal replied:
@glassez 

> Currently qBittorrent isn't supposed to be a set of several libraries.
The fact that CMake project divides it into several (static) libraries can only be considered as a workaround that allows CMake to follow the file structure of the project's source code (perhaps due to some drawbacks that prevent it from being done in another way).

When shared libraries are out of the question, the Professional CMake book still recommends building projects as a set of static libraries, as opposed to using plain `OBJECT` libraries, unless there is a special reason that requires its semantics.

> You should not change this here,

I did not change anything - like I said - `BUILD_SHARED_LIBS` is something that always exists, even if you don't do anything, and it's `OFF` by default. Like previously, the project is till built as a set of static libraries by default.

> especially since the build as a set of shared libraries is not possible currently (at least on Windows, it requires changes to the source code).

Yes, I know such a build is currently not possible, hence the "(for experimental development purposes)" part of the description. No one would normally turn the feature on.

The question becomes: do you think such a build with shared libraries will never be possible, or that it doesn't make sense to indicate that one day it might be? If so, it is a good idea to remove this. Otherwise, we can keep it in, it might encourage someone to investigate and submit patches to make it happen.

FranciscoPombal replied:
@glassez 

Well, to be honest I don't really care enough about this change to keep pushing for it right now. Its potential benefits for the near and medium term future are unclear, and if we do need such a change for the far future, its quite trivial anyway. So I reverted the internal libraries to be explicitly `STATIC`, like previously, and removed the `BUILD_SHARED_LIBS` description from the output.

glassez replied:
>So I reverted the internal libraries to be explicitly STATIC, like previously, and removed the BUILD_SHARED_LIBS description from the output.

:+1:

glassez commented on CMakeLists.txt:
```diff
@@ -1,71 +1,52 @@
+    feature_option(MSVC_RUNTIME_DYNAMIC "Use MSVC dynamic runtime library (-MD) instead of static (-MT)" OFF)
 endif()
 
+set(QBT_VER_STATUS "dev" CACHE STRING "Project status version. Should be empty for release builds.")
```
What is it for?

FranciscoPombal replied:
@glassez 

A replacement for the previous `alpha1` "status version" string (since we're no longer reading it from the `.pri` file). Like previously, it is appended to the project version, so that it becomes something like `4.3.0.0dev`. I might flesh this out further or remove it in the future refactor mentioned in https://github.com/qbittorrent/qBittorrent/pull/13327#discussion_r485144284.

glassez replied:
Anyway we should use the same value in both build systems.

FranciscoPombal replied:
@glassez Fixed (changed to `alpha1`). Although I would have preferred changing both to `dev`... but that can be a discussion for another time.

glassez commented on cmake/Modules/CompileFeature.cmake:
```diff
@@ -1,22 +0,0 @@
-# Helper function for coupling add_feature_info(), option(), and add_definitions()
```
Is it hard to extract this change into separate PR?
I mean removing this file and using `feature_option` instead of `optional_compile_definitions`?

FranciscoPombal replied:
@glassez 

Too complex to factor this out IMO, if you don't want it too be changed at all in this PR. It actually sets global compile definitions via directory-level properties which is very incompatible with the new architecture.

glassez commented on cmake/Modules/DeployQt5.cmake:
```diff
@@ -1,355 +0,0 @@
-# Borrowed from Avogadro project (https://github.com/OpenChemistry/avogadroapp)
```
Can't removing of this file and related changes be trivially extracted into separate PR?

FranciscoPombal replied:
@glassez 

IMO, if we keep it in this PR, we signal the intent that it is used/will be used. Since it is no longer used, from here on out, it is also removed at this point. It should definitely be its own separate commit though.

glassez replied:
>if we keep it in this PR, we signal the intent that it is used/will be used

When I suggest separating something from this PR into a separate one, I assume that it should be reviewed and merged before this one.

FranciscoPombal replied:
@glassez 

This change does not need to be further reviewed separately, _because it has already been reviewed_. Everyone agreed with this change in the previous version of this PR. The new way of creating the macOS bundle doesn't need this file at all. It has even been previously tested by Kolcha before and verified to work (you can also verify it yourself by inspecting the result of this build: https://github.com/FranciscoPombal/qBittorrent/actions/runs/223928661).

This file was only needed because of the `install_qt5_executable` function, which was previously only called once by `qBittorrent/cmake/Modules/bundle.cmake`, which in turn was a snippet of code called only once from here:

https://github.com/qbittorrent/qBittorrent/blob/defdd5121c940dfc3b807401338e15f66adfd671/src/app/CMakeLists.txt#L161

glassez commented on cmake/Modules/FunctionQbtCXXCompilerAndModeCheck.cmake:
```diff
@@ -1,24 +0,0 @@
-# Function for ensuring the build does not use a lower than the required minimum C++ mode, _min_std.
```
Isn't a good candidate for extracting?

FranciscoPombal replied:
@glassez same reasoning as  https://github.com/qbittorrent/qBittorrent/pull/13327#discussion_r485157109.

glassez commented on cmake/Modules/MacroConfigureMSVCRuntime.cmake:
```diff
@@ -1,36 +0,0 @@
-macro(configure_msvc_runtime)
```
Isn't a good candidate for extracting?
You can extract all similar (more or less) trivial cleanups into separate PR.

FranciscoPombal replied:
@glassez same reasoning as  https://github.com/qbittorrent/qBittorrent/pull/13327#discussion_r485157109.

glassez commented on src/app/CMakeLists.txt:
```diff
@@ -1,162 +1,184 @@
-    set_target_properties(qBittorrent PROPERTIES
-        MACOSX_BUNDLE_BUNDLE_NAME "${qbt_BUNDLE_NAME}"
-        MACOSX_BUNDLE_INFO_PLIST ${qBittorrent_BINARY_DIR}/dist/mac/Info.plist
+if (GUI AND ((CMAKE_SYSTEM_NAME STREQUAL "Windows") OR (CMAKE_SYSTEM_NAME STREQUAL "Darwin")))
```
I would still split Platform and Feature conditions even if it adds yet another indentation level.

FranciscoPombal replied:
@glassez fixed.

glassez commented on src/gui/CMakeLists.txt:
```diff
@@ -188,33 +186,9 @@ add_library(qbt_gui STATIC
-endif()
-
-target_include_directories(qbt_gui PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
+target_sources(qbt_gui INTERFACE about.qrc)
```
Please explain the changes like this one.

FranciscoPombal replied:
Similar reasoning to https://github.com/qbittorrent/qBittorrent/pull/13327#discussion_r485134005. `app` needs access to this. In this new architecture, if this is `PRIVATE`, you'll end up with missing/broken stuff in the `About` dialog.

glassez commented on src/gui/CMakeLists.txt:
```diff
@@ -223,13 +197,41 @@ target_link_libraries(qbt_gui
+            programupdater.cpp
+    )
+
+    if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
```
I would move this and the next `if`s outside of outer one.

FranciscoPombal replied:
@glassez fixed.

glassez reviewed (commented):
I finished my review.
Please don't add anything new here. Let's just fix the disadvantages and get it over with.

FranciscoPombal commented on src/webui/CMakeLists.txt:
```diff
@@ -1,38 +1,38 @@
 )
 
-qbt_target_sources(qBittorrent PRIVATE www/webui.qrc)
+target_sources(qbt_webui INTERFACE www/webui.qrc)
```
In this new implementation, it has to be `INTERFACE` because app needs access to it. If you try building it with `PRIVATE`, accessing the WebUI will just yield a web page saying "Not Found".

glassez replied:
> because app needs access to it

I see it. I just don't understand why did you change it. Why all web UI related stuff can't be built into corresponding library?
Well, maybe it's just that the splitting into libraries is still something inappropriate...

FranciscoPombal replied:
@glassez 

Actually, I just found out that for Qt, only the main application component needs to know about the resource sources, regardless of what component they are used in. So I have corrected the code to the following:

```diff
diff --git a/src/app/CMakeLists.txt b/src/app/CMakeLists.txt
index d01dfaeb4..567a617cb 100644
--- a/src/app/CMakeLists.txt
+++ b/src/app/CMakeLists.txt
@@ -140,6 +140,7 @@ endif()
 
 if (GUI)
     set_target_properties(qbt_app PROPERTIES OUTPUT_NAME qbittorrent)
+    target_sources(qbt_app PRIVATE "${qBittorrent_SOURCE_DIR}/src/gui/about.qrc")
     target_link_libraries(qbt_app PRIVATE qbt_gui)
 else()
     set_target_properties(qbt_app PROPERTIES OUTPUT_NAME qbittorrent-nox)
@@ -147,6 +148,7 @@ endif()
 
 if (WEBUI)
     target_sources(qbt_app PRIVATE
+        "${qBittorrent_SOURCE_DIR}/src/webui/www/webui.qrc"
         ${QBT_WEBUI_QM_FILES}
         ${qBittorrent_BINARY_DIR}/src/webui/www/translations/webui_translations.qrc # yes, it's supposed to be "*_BINARY_DIR"
     )
diff --git a/src/gui/CMakeLists.txt b/src/gui/CMakeLists.txt
index e8595ce2f..b8188f52c 100644
--- a/src/gui/CMakeLists.txt
+++ b/src/gui/CMakeLists.txt
@@ -188,8 +188,6 @@ add_library(qbt_gui STATIC
     updownratiodialog.ui
 )
 
-target_sources(qbt_gui INTERFACE about.qrc)
-
 target_link_libraries(qbt_gui
     PRIVATE
         qbt_base
diff --git a/src/webui/CMakeLists.txt b/src/webui/CMakeLists.txt
index 51cdfe489..a0d14a70f 100644
--- a/src/webui/CMakeLists.txt
+++ b/src/webui/CMakeLists.txt
@@ -33,6 +33,4 @@ add_library(qbt_webui STATIC
     webui.cpp
 )
 
-target_sources(qbt_webui INTERFACE www/webui.qrc)
-
 target_link_libraries(qbt_webui PRIVATE qbt_base)

```

glassez replied:
>Actually, I just found out that for Qt, only the main application component needs to know about the resource sources, regardless of what component they are used in

It doesn't seem quite logical that linking 
library is not enough to use it.

FranciscoPombal replied:
@glassez 

> It doesn't seem quite logical that linking
library is not enough to use it.

I don't know exactly why either, but that's not enough. Probably due to some Qt implementation detail.

glassez replied:
>I don't know exactly why either, but that's not enough. Probably due to some Qt implementation detail.

Then I would leave your original version to keep it where it comes from:
```cmake
target_sources(qbt_webui INTERFACE www/webui.qrc)
```
so using the library will only require one instruction, as expected:
```cmake
target_link_libraries(qbt_app PRIVATE qbt_webui)
```

FranciscoPombal replied:
@glassez ok, fixed.

FranciscoPombal commented on cmake/Modules/MacroQbtCommonConfig.cmake:
```diff
@@ -0,0 +1,117 @@
+                NOMINMAX
+        )
+    elseif (UNIX AND (NOT APPLE) AND (NOT CYGWIN))
+        target_compile_definitions(qbt_common_cfg INTERFACE _DEFAULT_SOURCE)
```
@glassez 

I mostly left it there because it was there previously. It's a glibc thing, I think it's "common practice" to define it. Otherwise, gcc can complain about some implicit function declartions, if we use those certain functions.

There is probably no problem in defining it in other systems I guess (they probably just ignore it). I wonder if it's a problem for Linux systems using something like `musl` though.

At the moment, we don't need it - I just tried building without it and got no warnings. Should I remove it? I think this is mostly for C stuff anyway, not C++.

More info: 

https://stackoverflow.com/questions/29201515/what-does-d-default-source-do (also read the second answer)
https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html.

glassez replied:
At least you shouldn't separate different Unix systems according to it. Let either declare or remove it everywhere.

FranciscoPombal replied:
@glassez Well, since you have no reasons for keeping it either, I've removed it.

glassez replied:
>@glassez Well, since you have no reasons for keeping it either, I've removed it.

Don't count on me here. I'm not sure what it is, why it's needed, or whether we can safely remove it.

FranciscoPombal replied:
@glassez To the best of my knowledge right now, I think it is safe to remove it.

FranciscoPombal commented on CMakeLists.txt:
```diff
@@ -1,71 +1,52 @@
-include(GNUInstallDirs)
-include(FeatureSummary)
+project(qBittorrent
+    VERSION 4.3.0.0
```
@glassez 

Isn't this meant to replace qmake in the somewhat near future? I made this change because I didn't think it was a good idea to always read this from the `.pri` and therefore have the CMake buildsystem implicitly depending on the qmake one. In fact, it's not even a good idea to pass the project version as a compile definition, since it causes recompilation of everything. Having it defined this way is the first step towards a minor version number-related refactor I have planned for the future that will solve this problem.

In the meantime, it's one more number to change in commits that increment the version - not much more work, and I'll make sure to nag about it if anyone forgets.

glassez replied:
>Isn't this meant to replace qmake in the somewhat near future? 

I think it's more about a few years.
Most likely, we should make both systems equally "official" with the start of Qt6 support. So I would not want edit several different files to change the project version, which can cause a lack of synchronization.

FranciscoPombal replied:
@glassez 

> So I would not want edit several different files to change the project version

This is simply the price to pay for maintaining 2 different build systems like this. But it's not a very high one - it's just a single additional line change, and not very often. We're much worse off if we couple parts of one buildsystem's logic to the other IMO.

>I think it's more about a few years.
Most likely, we should make both systems equally "official" with the start of Qt6 support.

Oh god...

glassez replied:
>This is simply the price to pay for maintaining 2 different build systems like this. But it's not a very high one - it's just a single additional line change, and not very often

Okay, it's not that much of a problem for me if others agree with you.

FranciscoPombal commented on src/app/CMakeLists.txt:
```diff
@@ -1,162 +1,184 @@
+# Executable target configuration
+# -----------------------------------------------------------------------------
+# -----------------------------------------------------------------------------
+add_executable(qbt_app)
```
@glassez I changed it for good reason - to not clash with the project name (the name passed in the call to `PROJECT()`). Otherwise it is confusing and there may be surprises, because CMake defines variables based on the project name.

FranciscoPombal commented:
@glassez 

A note about the review comments about "extracting changes to other PRs" - I've explained in https://github.com/qbittorrent/qBittorrent/pull/13327#discussion_r485157109 why I don't think it's a good idea. However, I do agree those should be separated into distinct _commits_. I'll do that once we get over the first review.

FranciscoPombal commented on dist/CMakeLists.txt:
```diff
@@ -1,9 +1,7 @@
     add_subdirectory(mac)
-elseif (UNIX)
+elseif (UNIX AND (NOT APPLE) AND (NOT CYGWIN))
     add_subdirectory(unix)
```
@glassez Not sure, now that I really think about it. What do you think? Should I enable it for Cygwin?

glassez replied:
>Should I enable it for Cygwin?

I think Yes.

FranciscoPombal replied:
@glassez fixed.

FranciscoPombal commented on cmake/Modules/MacroQbtCommonConfig.cmake:
```diff
@@ -0,0 +1,117 @@
+        target_compile_definitions(qbt_common_cfg INTERFACE _DEFAULT_SOURCE)
+    endif()
+
+
```
@glassez fixed.

FranciscoPombal commented on cmake/Modules/MacroQbtCommonConfig.cmake:
```diff
@@ -0,0 +1,117 @@
+        # Clang 11 still doesn't support -Wstrict-null-sentinel
+        include(CheckCXXCompilerFlag)
+        check_cxx_compiler_flag(-Wstrict-null-sentinel SNS_SUPPORT)
+        if(SNS_SUPPORT)
```
@glassez fixed.

FranciscoPombal commented on cmake/Modules/MacroQbtCommonConfig.cmake:
```diff
@@ -0,0 +1,117 @@
+    #    set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
+    #endif()
+
+    # There is no generator expression for CMAKE_GENERATOR yet
```
@glassez fixed.

FranciscoPombal commented on cmake/Modules/MacroQbtCommonConfig.cmake:
```diff
@@ -0,0 +1,117 @@
+    # build type-dependent functionality for both single config and multi-config-generators alike
+    set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>$<$<BOOL:${MSVC_RUNTIME_DYNAMIC}>:DLL>")
+    # GENEXP: Alternatively, to _really_ minimize genexp use...
+    #if (MSVC_RUNTIME_DYNAMIC)
```
@glassez fixed.

FranciscoPombal commented on src/app/CMakeLists.txt:
```diff
@@ -1,162 +1,184 @@
+add_executable(qbt_app)
+
+target_sources(qbt_app
+    PRIVATE
```
@glassez fixed, and fixed all such cases of excessive indentation.

FranciscoPombal commented on src/app/CMakeLists.txt:
```diff
@@ -1,162 +1,184 @@
-# translations
-include(QbtTranslations)
+target_link_libraries(qbt_app
+    PRIVATE
```
@glassez fixed.

FranciscoPombal commented on src/app/CMakeLists.txt:
```diff
@@ -1,162 +1,184 @@
+    )
+    set_target_properties(qbt_app
+        PROPERTIES
+            MACOSX_BUNDLE             ON
```
@glassez fixed.

FranciscoPombal commented on src/app/CMakeLists.txt:
```diff
@@ -1,162 +1,184 @@
+    )
+    # misc macOS stuff
+    target_sources(qbt_app
+        PRIVATE
```
@glassez fixed.

FranciscoPombal commented on src/CMakeLists.txt:
```diff
@@ -1,62 +1,75 @@
+            TYPE REQUIRED
+        )
+    else()
+        set_package_properties(LibtorrentRasterbar PROPERTIES TYPE REQUIRED)
```
@glassez 

If the package is found without pkg-config, we set the `TYPE` to `REQUIRED`, so that it displays correctly as a required package. This is all due to the fact that on Linux, we first find libtorrent as a non-`REQUIRED` package to give it a second chance with pkg-config.

FranciscoPombal commented:
@glassez I've addressed the comments. The ones I did not mark as resolved are the ones that I think warrant further clarifications/discussion. Once we iron those out I'll force push to remove the `# GENEXP:` comments.

FranciscoPombal commented:
@glassez Ok, I think all comments of your comments of the first review have been addressed at this point. Let me know if there is something you think needs to be discussed further though. Otherwise, I'll remove the `# GENEXP:` comments and then start splitting this up into separate commits.

glassez commented:
>Ok, I think all comments of your comments of the first review have been addressed at this point. Let me know if there is something you think needs to be discussed further though.

Seems it has no more blocking issues for me. Although I need to review it again after all final changes have done.

FranciscoPombal commented:
> Seems it has no more blocking issues for me. Although I need to review it again after all final changes have done.

Alright then. @Chocobo1 please take a look at this. The changes made as a result of @glassez's review comments were still very minor, so you can still compare this in its current state directly against https://github.com/qbittorrent/qBittorrent/pull/12746 to ascertain the value/advantages of going the generator expressions route.

Again, I will note that there are a few comments starting with `# GENEXP:` to signal places where the use of generator expressions was unavoidable. Once I address your review comments, I can remove those comments and finally start splitting this up into discrete commits - I already have a pretty good idea of how I will do that; @glassez also threw in some good suggestions.

FranciscoPombal commented:
Force push to resolve https://github.com/qbittorrent/qBittorrent/pull/13327#discussion_r485750344

Chocobo1 commented on cmake/Modules/MacroQbtCommonConfig.cmake:
```diff
@@ -0,0 +1,106 @@
+
+    add_library(qbt_common_cfg INTERFACE)
+
+    # Full C++ 14 support is required
```
How do you decide to put what features in the list? I mean is there a compiled list or something like that?

FranciscoPombal replied:
@Chocobo1 

> How do you decide what to put in the list?

I went through https://cmake.org/cmake/help/v3.16/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#individual-features-from-c-14 and put in all those that had to do with C++14.

> I mean is there a compiled list or something like that?

Recently I submitted a PR to improve CMake's documentation in this regard: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/5151/diffs.
The new result can be seen here, with features properly sorted by C++ standard:

https://cmake.org/cmake/help/git-master/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#high-level-meta-features-indicating-c-standard-support
https://cmake.org/cmake/help/git-master/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#low-level-individual-compile-features
https://cmake.org/cmake/help/git-master/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#individual-features-from-c-14

FranciscoPombal replied:
@Chocobo1 Just after typing this, I reviewed the sorted list in https://cmake.org/cmake/help/git-master/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#individual-features-from-c-14 and realized I actually missed a couple. Also I took the opportunity to sort them all alphabetically after `cxx_std_14`. Fixed.

Chocobo1 replied:
>Just after typing this, I reviewed the sorted list in https://cmake.org/cmake/help/git-master/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#individual-features-from-c-14 and realized I actually missed a couple. Also I took the opportunity to sort them all alphabetically after cxx_std_14. Fixed.

You should add an URL link to the cmake document.

FranciscoPombal replied:
fixed.

Chocobo1 commented on cmake/Modules/MacroQbtCommonConfig.cmake:
```diff
@@ -0,0 +1,106 @@
+        )
+    endif()
+
+    if((CXX_COMPILER_ID STREQUAL "GNU") OR (CXX_COMPILER_ID STREQUAL "Clang") OR (CXX_COMPILER_ID STREQUAL "AppleClang"))
```
`if (` and please check globally.

FranciscoPombal replied:
fixed all instances.

Chocobo1 commented on src/app/CMakeLists.txt:
```diff
@@ -15,148 +43,136 @@ add_executable(qBittorrent
-    find_package(Qt5Svg REQUIRED)
-    include(bundle)
+if (MSVC)
+    # GENEXP: honestly don't know how to do this without genexp
```
Please remove this kind of comments. Those are only suitable on github reviews.

FranciscoPombal replied:
fixed.

Chocobo1 commented on cmake/Modules/MacroQbtCommonConfig.cmake:
```diff
@@ -0,0 +1,106 @@
+            _UNICODE
+            WIN32_LEAN_AND_MEAN
+            _CRT_SECURE_NO_DEPRECATE
+            _SCL_SECURE_NO_DEPRECATE
```
Can you try compiling without the above 2 macros? If removing them does not produce additional compile warnings/errors then please remove them, as they are not present in winconf.pri either.

FranciscoPombal replied:
Removed, compiled with no errors or warnings.

I also learned the following:

- It is not necessary to specify the value of `_WIN32_IE`, so I removed that, too. It is automatically set to the correct value in the `sdkddkver.h` Windows header based on the version set in `_WIN32_WINNT`.

- With `_WIN32_WINNT` defined, it is not necessary to define `WIN_VER`, it gets set automatically in the same header.

- `NTDDI_VERSION` is useful for finer grained version control, but currently it does nothing, since there is no finer grain control possibie for Windows 7 versions. It will come in handy when we need to distinguish among Windows 10 versions.

References:
https://docs.microsoft.com/en-us/windows/win32/WinProg/using-the-windows-headers
https://docs.microsoft.com/en-us/cpp/porting/modifying-winver-and-win32-winnt

- `UNICODE` and `_UNICODE` should both be defined (as they are now)

References:
https://docs.microsoft.com/en-us/cpp/text/unicode-programming-summary
https://docs.microsoft.com/en-us/windows/win32/learnwin32/working-with-strings

> Be careful: Some headers use the preprocessor symbol `UNICODE`, others use `_UNICODE` with an underscore prefix. Always define both symbols. Visual C++ sets them both by default when you create a new project.

Chocobo1 replied:
>Removed, compiled with no errors or warnings.

OK.

>It is not necessary to specify the value of _WIN32_IE, so I removed that, too. It is automatically set to the correct value in the sdkddkver.h Windows header based on the version set in _WIN32_WINNT.

You shouldn't do more changes than asked here. Now you deviate from winconf.pri again, if you want `_WIN32_IE` (or anything) removed, you should proposed in another PR.

FranciscoPombal replied:
@Chocobo1 Alright, fixed.

FranciscoPombal commented:
@Chocobo1 Force-pushed to address most review comments.

I will now investigate the situation with `_CRT_SECURE_NO_DEPRECATE` and `_SCL_SECURE_NO_DEPRECATE`.

FranciscoPombal commented:
Seems like `_CRT_SECURE_NO_DEPRECATE` and `_SCL_SECURE_NO_DEPRECATE` have been replaced by `_CRT_SECURE_NO_WARNINGS` and `_SCL_SECURE_NO_WARNINGS`, respectively, for a long time. Apparently they just disable a `/W3` deprecation warning (`C4996`) for certain things that we probably ought to have enabled in all cases anyway.

https://docs.microsoft.com/en-us/cpp/standard-library/safe-libraries-cpp-standard-library
https://docs.microsoft.com/en-us/cpp/c-runtime-library/security-features-in-the-crt
https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4996

I'm not even sure if they were having any effect until now. I'll try building without them soon to know for sure.

FranciscoPombal commented:
@Chocobo1 All comments addressed.

Chocobo1 reviewed (approved):
@glassez 
I finished my review, you can take a final look.

glassez commented:
@FranciscoPombal
Just to clarify one point. Previously, I suggested splitting these changes into some separate parts, but my goal was to review and merge them one by one (as separate PRs). Since it is generally approved now, I no longer see the point of making such a splitting post factum. The current split into two commits doesn't make sense for the master git history either. So let's just merge both commits into one and finish with it (if you don't have a great desire to prolong the fun even more).

FranciscoPombal commented:
@glassez 

>Just to clarify one point. Previously, I suggested splitting these changes into some separate parts, but my goal was to review and merge them one by one (as separate PRs). Since it is generally approved now, I no longer see the point of making such a splitting post factum.

OK.

>The current split into two commits doesn't make sense for the master git history either. So let's just merge both commits into one and finish with it (if you don't have a great desire to prolong the fun even more).

The purpose of the current commit split is to enable easy comparison against the alternative with generator expressions, without having to open up https://github.com/qbittorrent/qBittorrent/pull/12746:

>It is built directly on top of it, to enable easily studying the differences from the diff, both against master and the genexp alternative.

Now that the general review is done, and both implementations have been verified to produce the same results, they can be compared fairly in terms of readability, maintainability, etc (actually, the genexp version is missing a couple minor fixes that came in during this review, like removing the `*_SECURE_NO_DEPRECATE` defines for Windows, which doesn't really make a difference - but let me know if you want me to remove that noise). I await your comments/decision on this matter (@Chocobo1 you too).

glassez commented:
@FranciscoPombal
I don't think we should stay here any longer. If you still want to present genexps as an improvement, you can publish the corresponding PR on top of this one when it is merged.

FranciscoPombal commented:
@glassez 

> @FranciscoPombal
> I don't think we should stay here any longer. If you still want to present genexps as an improvement, you can publish the corresponding PR on top of this one when it is merged.

I was under the impression that there would be some more discussion about this before merging either one, but that's fine, the existing improvements can be merged now and we'll handle the rest later. I force-pushed to squash the commits together. This is ready for merging now.

glassez commented:
Thank you!

NotTsunami commented:
All CMake builds on CI are failing now, with osx failing in a previous spot from before.

Chocobo1 commented:
>All CMake builds on CI are failing now, with osx failing in a previous spot from before.

Linux failing point:
Before this PR: No error
This PR: https://travis-ci.org/github/qbittorrent/qBittorrent/jobs/728209129#L1579

OSX failing point:
Before this PR: https://travis-ci.org/github/qbittorrent/qBittorrent/jobs/727897569#L2701
This PR: https://travis-ci.org/github/qbittorrent/qBittorrent/jobs/728209131#L1225

It should be investigated.

FranciscoPombal commented:
@Chocobo1 @NotTsunami 

> > All CMake builds on CI are failing now, with osx failing in a previous spot from before.
> 
> Linux failing point:
> Before this PR: No error
> This PR: https://travis-ci.org/github/qbittorrent/qBittorrent/jobs/728209129#L1579

I have also seen this (or very similar) reported here: https://github.com/qbittorrent/qBittorrent/issues/13347 
This failure seems to affect some systems, but not others. I don't know the exact reason why (and I can't reproduce on my Linux and Windows machines, though I haven't tried with WSL, like in the case of the bug report). I think I think this would fix it, but I wanted confirmation before proceeding: https://github.com/qbittorrent/qBittorrent/issues/13347#issuecomment-689887545 It does seem right to add the explicit direct dependency, as we do use it in `src/base/utils/misc.cpp` and `src/base/utils/password.cpp`. Actually, I can just submit a PR with the change and see what happens with CI. Will do that.

> OSX failing point:
> Before this PR: https://travis-ci.org/github/qbittorrent/qBittorrent/jobs/727897569#L2701
> This PR: https://travis-ci.org/github/qbittorrent/qBittorrent/jobs/728209131#L1225

Though the symptom is different, he reason is the same - the broken bottle currently used. https://github.com/qbittorrent/qBittorrent/pull/13288 solves this of course.

FranciscoPombal commented:
@Chocobo1 @NotTsunami 

https://github.com/qbittorrent/qBittorrent/pull/13402
